Sources/Tools/plFilePatcher/plFilePatcher.cpp

```diff
@@ -0,0 +1,186 @@
+/*==LICENSE==*
+
+CyanWorlds.com Engine - MMOG client, server and tools
+Copyright (C) 2011  Cyan Worlds, Inc.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "plFilePatcher.h"
+
+#include <string_theory/format>
+
+#include "pnAsyncCore/pnAsyncCore.h"
+#include "plNetGameLib/plNetGameLib.h"
+#include "pfConsoleCore/pfConsoleEngine.h"
+#include "pfPatcher/plManifests.h"
+
+const int kNetTransTimeout = 5 * 60 * 1000;                     // 5m
+const int kAsyncCoreShutdownTime = 2 * 1000;                    // 2s
+const std::chrono::milliseconds kNetCoreUpdateSleepTime(10);    // 10ms
+
+plFilePatcher::plFilePatcher(const plFileName& serverIni)
+    : fFlags(kPatchEverything), fServerIni(serverIni), fError(),
+    fNetCoreState(kNetCoreInactive), fProgressFunc(), fDownloadFunc()
+{
+}
+
+PF_CONSOLE_LINK_FILE(Core)
+bool plFilePatcher::ILoadServerIni()
+{
+    PF_CONSOLE_INITIALIZE(Core);
+
+    pfConsoleEngine console;
+    return console.ExecuteFile(fServerIni);
+}
+
+void plFilePatcher::IInitNetCore()
+{
+    AsyncCoreInitialize();
+
+    NetClientInitialize();
+    NetClientSetErrorHandler([](ENetProtocol protocol, ENetError error) {
+        //ST::printf("Protocol: {} -- {}\n", NetProtocolToString(protocol), NetErrorAsString(error));
```

Comment:
Commented code 😢

Reply:
I need to figure out how to deal with this, since I can't seem to pass in a lambda or `std::bind` function here that preserves the context of the plFilePatcher 😒 
I'd like to be able to do `ISetNetError(msg)` here and `ISignalNetCoreShutdown()`

Reply:
Yeah, you'll have to pass in a function pointer - lambdas generally have trouble casting to function pointers. The `plFilePatcher` instance will probably need to be a singleton because the function pointer doesn't let you receive an optional parameter. Sad.

Reply:
Is it feasible to change the netcode to take a std::function? That should allow a function pointer or a bound function, no?